server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -195,13 +214,8 @@ private void collectAndRecordNodeWeightStats(Balancer balancer, WeightFunction w
 
     @Override
     public ShardAllocationDecision decideShardAllocation(final ShardRouting shard, final RoutingAllocation allocation) {
-        WeightFunction weightFunction = new WeightFunction(
-            shardBalanceFactor,
-            indexBalanceFactor,
-            writeLoadBalanceFactor,
-            diskUsageBalanceFactor
-        );
-        Balancer balancer = new Balancer(writeLoadForecaster, allocation, weightFunction, threshold);
+        WeightFunction weightFunction = new SpecialisedWeightFunction(settings);
```

Comment:
If we wanted to we could have `WeightFunctionFactory` implementations for stateful and stateless (the stateful one just returns a `SingleWeightFunction` with the non-specific weights, and the stateless one returns a `SpecialisedWeightFunction`).